v2/pkg/delete/delete_images.go

```diff
@@ -0,0 +1,297 @@
+	}
+	return images, nil
+}
+
+// CollectReleaseImages
+func (o DeleteImages) CollectReleaseImages(releaseFolder string) ([]v1alpha3.CopyImageSchema, error) {
+	var rs v1alpha3.ReleaseSchema
+	releaseJson := filepath.Join(releaseFolder, releaseManifests, imageReferences, "/")
+	data, err := os.ReadFile(releaseJson)
+	if err != nil {
+		return []v1alpha3.CopyImageSchema{}, err
+	}
+	err = json.Unmarshal(data, &rs)
+	if err != nil {
+		return []v1alpha3.CopyImageSchema{}, err
+	}
+
+	// collect all release images and add them to CopyImageSchema collection
+	var allImages []v1alpha3.CopyImageSchema
+	for _, img := range rs.Spec.Tags {
+		copyIS, err := buildFormatedCopyImageSchema(img.Name, img.From.Name, o.LocalStorageFQDN)
+		if err != nil {
+			return []v1alpha3.CopyImageSchema{}, err
+		}
+		allImages = append(allImages, copyIS)
+	}
+	return allImages, nil
+}
+
+// CollectOperatorImages
+func (o DeleteImages) CollectOperatorImages() ([]v1alpha3.CopyImageSchema, error) {
+	var allImages []v1alpha3.CopyImageSchema
+	for _, op := range o.Config.Mirror.Operators {
+		imageIndexDir := filepath.Base(op.Catalog)
+		cacheDir := strings.Join([]string{o.Opts.Global.WorkingDir, operatorImageExtractDir, strings.ReplaceAll(imageIndexDir, ":", "/")}, "/")
+		// we dont know the subfolder name so lets get it
+		dir, err := os.ReadDir(cacheDir)
+		if err != nil {
+			return nil, err
+		}
+		if len(dir) > 0 {
+			// if there is more than one dir there could be a problem
+			// for now we select the first one
+			o.Log.Debug("label (directory) %s", dir[0].Name())
+			operatorCatalog, err := o.Manifest.GetCatalog(filepath.Join(cacheDir, dir[0].Name()))
+			if err != nil {
+				return []v1alpha3.CopyImageSchema{}, err
+			}
+			ri, err := o.Manifest.GetRelatedImagesFromCatalog(operatorCatalog, op)
+			if err != nil {
+				return []v1alpha3.CopyImageSchema{}, err
+			}
+			// collect all operator images and add them to CopyImageSchema collection
+			for _, v := range ri {
+				for _, i := range v {
+					copyIS, err := buildFormatedCopyImageSchema(i.Name, i.Image, o.LocalStorageFQDN)
+					if err != nil {
+						o.Log.Error(deleteImagesErrMsg, err)
+					}
+					allImages = append(allImages, copyIS)
+				}
+			}
+		} else {
+			return nil, fmt.Errorf("no (label) directory found in %s", cacheDir)
+		}
+	}
+	return allImages, nil
+}
+
+// CollectAdditionalImages
+func (o DeleteImages) CollectAdditionalImages() ([]v1alpha3.CopyImageSchema, error) {
+	var allImages []v1alpha3.CopyImageSchema
+	for _, ai := range o.Config.Mirror.AdditionalImages {
+		is, err := buildFormatedCopyImageSchema(ai.Name, ai.Name, o.LocalStorageFQDN)
+		if err != nil {
+			o.Log.Error(deleteImagesErrMsg, err)
+		}
+
+		allImages = append(allImages, is)
+	}
+	return allImages, nil
+}
+
+// buildFormatedCopyImageSchema - simple private utility to build the CopyImageSchema data
+func buildFormatedCopyImageSchema(name, img, localStorageFQDN string) (v1alpha3.CopyImageSchema, error) {
```

Comment:
Suggestion: would it not be easier if `buildFormatedCopyImageSchema` followed what `prepareD2MCopyBatch` does for other collectors?
The idea would be that the CopyImageSchema.Source would contain a reference to the local cache image, while the CopyImageSchema.Destination would contain the remote registry image reference. 
This way, there is no need for extra Split and Join logic in the DeleteCacheBlobs or DeleteRegistryImages I believe.

Overall the logic in this file repeats the one in the various collectors. I don't know if there is an opportunity to avoid this duplication, maybe by splitting more the functions in the ReleaseImageCollector and OperatorImageCollector, so that we can reuse the code in multiple places? 
BTW, it was already discussed that the prepareM2D and prepareD2M is somehow duplicated between the ReleaseImageCollector and OperatorImageCollector, and we needed to rewrite this better. 
In case this can be achieved, the benefit would be the ease of fixing bugs for both the delete and the mirror workflows in one shot.

Reply:
Yes let me  see if I can simplify this